File: flink-datastream-1.7.1/src/main/scala/scalalang/source/SensorSource.scala
package scalalang.source

import java.time.Instant

import org.apache.flink.streaming.api.functions.source.{RichParallelSourceFunction, SourceFunction}
import scalalang.util.SensorReading

import scala.util.Random

class SensorSource extends RichParallelSourceFunction[SensorReading] {
  var running = true

  override def run(sourceContext: SourceFunction.SourceContext[SensorReading]): Unit = {
    val rand = new Random()
    val taskIdx = this.getRuntimeContext.getIndexOfThisSubtask
    var curFTemp = (1 to 10).map(
      i => ("sensor_" + (taskIdx * 10 + i), 65 + (rand.nextGaussian() * 20)))
    while (running) {
      curFTemp = curFTemp.map(t => (t._1, t._2 + (rand.nextGaussian() * 0.5)))
      //Calendar已经过时，改用Instant
      //val curTime = Calendar.getInstance().getTimeInMillis
      val curTime = Instant.now.toEpochMilli
      curFTemp.foreach(t => sourceContext.collect(SensorReading(t._1, curTime, t._2)))
      Thread.sleep(100)
    }
  }

  override def cancel(): Unit = {
    running = false
  }
}

File: flink-datastream-1.11.2/src/main/java/datastream/source/SensorSource.java
package datastream.source;

import datastream.util.SensorReading;
import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;

import java.time.Instant;
import java.util.Random;

public class SensorSource extends RichParallelSourceFunction<SensorReading> {
    private boolean running = true;
    @Override
    public void run(SourceContext<SensorReading> sourceContext) throws Exception {
        Random rand = new Random();
        int taskIdx = this.getRuntimeContext().getIndexOfThisSubtask();
        String[] sensorIds = new String[10];
        double[] curFTemp = new double[10];
        for (int i = 0; i < 10; i++) {
            sensorIds[i] = "sensor_" + (taskIdx * 10 + i);
            curFTemp[i] = 65 + (rand.nextGaussian() * 20);
        }
        while (running) {
            long curTime = Instant.now().toEpochMilli();
            for (int i = 0; i < 10; i++) {
                curFTemp[i] += rand.nextGaussian() * 0.5;
                sourceContext.collect(new SensorReading(sensorIds[i], curTime, curFTemp[i]));
            }
            Thread.sleep(100);
        }
    }
    @Override
    public void cancel() {
        this.running = false;
    }
}
